ledger/eval.go

```diff
@@ -1088,6 +1164,56 @@ func (eval *BlockEvaluator) finalValidation() error {
 				return fmt.Errorf("CompactCertType %d unexpected", ccType)
 			}
 		}
+
+		proto, ok := config.Consensus[eval.block.BlockHeader.CurrentProtocol]
+		if !ok {
+			return fmt.Errorf("unknown consensus version: %s", eval.block.BlockHeader.CurrentProtocol)
+		}
+
+		expectedMaxNumberOfExpiredAccounts := proto.MaxExpiredAccountsToProcess
+		lengthOfExpiredParticipationAccounts := len(eval.block.ParticipationUpdates.ExpiredParticipationAccounts)
+
+		// If the length of the array is strictly greater than our max then we have an error.
+		// This works when the expected number of accounts is zero (i.e. it is disabled) as well
+		if lengthOfExpiredParticipationAccounts > expectedMaxNumberOfExpiredAccounts {
+			return fmt.Errorf("length of expired accounts (%d) was greater than expected (%d)",
+				lengthOfExpiredParticipationAccounts, expectedMaxNumberOfExpiredAccounts)
+		}
+
+		// For security reasons, we need to make sure that all addresses in the expired participation accounts
+		// are unique.  We make this map to keep track of previously seen address
+		addressSet := make(map[basics.Address]bool)
+
+		// Validate that all expired accounts meet the current criteria
+		currentRound := eval.Round()
+		for _, accountAddr := range eval.block.ParticipationUpdates.ExpiredParticipationAccounts {
+
+			if _, exists := addressSet[accountAddr]; exists {
+				// We shouldn't have duplicate addresses...
+				return fmt.Errorf("duplicate address found: %v", accountAddr)
+			}
+
+			// Record that we have seen this address
+			addressSet[accountAddr] = true
+
+			acctData, err := eval.state.lookup(accountAddr)
+			if err != nil {
+				return fmt.Errorf("ExpiredParticipationAccount wrong: %v was not found", accountAddr)
```

Comment:
same note as before; just one more comment - errors should always start with a lower-case.